src/hardware/mixer.cpp

```diff
@@ -406,160 +405,169 @@ void MixerChannel::AddSamples(uint16_t len, const Type *data)
-							} else {
-								next_sample[0] = static_cast<int>(host_readd((HostPt)&data[pos * 2 + 0])) - 32768;
-								next_sample[1] = static_cast<int>(host_readd((HostPt)&data[pos * 2 + 1])) - 32768;
-							}
+							next_sample[0] = static_cast<int>(host_readd((HostPt)&data[pos * 2 + 0])) - offset;
+							next_sample[1] = static_cast<int>(host_readd((HostPt)&data[pos * 2 + 1])) - offset;
 						}
+					}
+				} else { // mono
+					if (nativeorder) {
+						next_sample[0] = static_cast<int>(data[pos]) - offset;
 					} else {
-						if (nativeorder) {
-							next_sample[0] = static_cast<int>(data[pos]) - 32768;
+						if (sizeof(Type) == 2) {
+							next_sample[0] = static_cast<int>(host_readw((HostPt)&data[pos])) - offset;
 						} else {
-							if ( sizeof( Type) == 2) {
-								next_sample[0] = static_cast<int>(host_readw((HostPt)&data[pos])) - 32768;
-							} else {
-								next_sample[0] = static_cast<int>(host_readd((HostPt)&data[pos])) - 32768;
-							}
+							next_sample[0] = static_cast<int>(host_readd((HostPt)&data[pos])) - offset;
 						}
 					}
 				}
 			}
-			//This sample has been handled now, increase position
-			pos++;
-#if MIXER_UPRAMP_STEPS > 0
-			if (last_samples_were_silence && pos == 1) {
-				offset[0] = next_sample[0] - prev_sample[0];
-				if (stereo) offset[1] = next_sample[1] - prev_sample[1];
-				//Don't bother with small steps.
-				if (offset[0] < (MIXER_UPRAMP_SAVE*4) && offset[0] > (-MIXER_UPRAMP_SAVE*4)) offset[0] = 0;
-				if (offset[1] < (MIXER_UPRAMP_SAVE*4) && offset[1] > (-MIXER_UPRAMP_SAVE*4)) offset[1] = 0;
-			}
-
-			if (offset[0] || offset[1]) {
-				next_sample[0] = next_sample[0] - (offset[0]*(MIXER_UPRAMP_STEPS*static_cast<Bits>(len)-static_cast<Bits>(pos))) /( MIXER_UPRAMP_STEPS*static_cast<Bits>(len) );
-				next_sample[1] = next_sample[1] - (offset[1]*(MIXER_UPRAMP_STEPS*static_cast<Bits>(len)-static_cast<Bits>(pos))) /( MIXER_UPRAMP_STEPS*static_cast<Bits>(len) );
-			}
-#endif
 		}
 
 		// Process initial samples through an expanding envelope to
 		// prevent severe clicks and pops. Becomes a no-op when done.
-		envelope.Process(stereo, interpolate, prev_sample, next_sample);
+		envelope.Process(stereo, prev_sample);
 
-		//Where to write
-		mixpos &= MIXER_BUFMASK;
-		if (!interpolate) {
+		const auto left = prev_sample[mapped_channel_left] * volmul[0];
+		const auto right = (stereo ? prev_sample[mapped_channel_right]
+		                           : prev_sample[mapped_channel_left]) *
+		                   volmul[1];
 
-			mixer.work[mixpos][mapped_output_left] +=
-			        prev_sample[mapped_channel_left] * volmul[0];
+		out_frame = {0.0f, 0.0f};
+		out_frame[mapped_output_left] += left;
+		out_frame[mapped_output_right] += right;
 
-			mixer.work[mixpos][mapped_output_right] +=
-			        (stereo ? prev_sample[mapped_channel_right]
-			                : prev_sample[mapped_channel_left]) *
-			        volmul[1];
+		out.push_back(out_frame[0]);
+		out.push_back(out_frame[1]);
```

Comment:
`emplace_back` be used, and skips creating a default in the vector first and then assigning.
(https://stackoverflow.com/questions/23717151/why-emplace-back-is-faster-than-push-back, even for a simply `int` type)

Reply:
Great, didn't know that!